cdap-app-fabric/src/main/java/io/cdap/cdap/internal/operations/OperationRunsStore.java

```diff
@@ -0,0 +1,384 @@
+    Range.Bound startBound = Range.Bound.INCLUSIVE;
+    Range.Bound endBound = Range.Bound.INCLUSIVE;
+    Collection<Field<?>> startFields = new ArrayList<>();
+
+    startFields.add(Fields.stringField(StoreDefinition.OperationRunsStore.NAMESPACE_FIELD,
+        request.getNamespace()));
+
+    Collection<Field<?>> endFields = startFields;
+
+    if (request.getScanFrom() != null) {
+      startBound = Range.Bound.EXCLUSIVE;
+      startFields = getRangeFields(request.getNamespace(), request.getScanFrom());
+    }
+    if (request.getScanTo() != null) {
+      endBound = Range.Bound.EXCLUSIVE;
+      endFields = getRangeFields(request.getNamespace(), request.getScanTo());
+    }
+
+    Range range = Range.create(endFields, endBound, startFields, startBound);
+    Collection<Field<?>> filterIndexes = getFilterIndexes(request.getFilter());
+    int limit = request.getLimit();
+
+    StructuredTable table = getOperationRunsTable(context);
+    try (CloseableIterator<StructuredRow> iterator = table.scan(range, limit,
+        filterIndexes, StoreDefinition.OperationRunsStore.START_TIME_FIELD, SortOrder.DESC)) {
+      boolean keepScanning = true;
+      while (iterator.hasNext() && keepScanning && limit > 0) {
+        StructuredRow row = iterator.next();
+        Map.Entry<String, OperationRunDetail<?>> scanEntry =
+            new SimpleImmutableEntry<String, OperationRunDetail<?>>(
+                row.getString(StoreDefinition.OperationRunsStore.ID_FIELD),
+                GSON.fromJson(row.getString(StoreDefinition.OperationRunsStore.DETAILS_FIELD),
+                    OperationRunDetail.class)
+            );
+        keepScanning = func.apply(scanEntry);
+        limit--;
+      }
+    }
+  }
+
+  private Collection<Field<?>> getFilterIndexes(OperationRunFilter filter) {
+    Collection<Field<?>> filterIndexes = new ArrayList<>();
+    if (filter == null) {
+      return filterIndexes;
+    }
+    if (filter.getOperationType() != null) {
+      filterIndexes.add(Fields.stringField(StoreDefinition.OperationRunsStore.TYPE_FIELD,
+          filter.getOperationType()));
+    }
+    if (filter.getStatus() != null) {
+      filterIndexes.add(Fields.stringField(StoreDefinition.OperationRunsStore.STATUS_FIELD,
+          filter.getStatus().toString()));
+    }
+    return filterIndexes;
+  }
+
+  private List<Field<?>> getRangeFields(String namespace, String id)
+      throws IOException, OperationRunNotFoundException {
+    List<Field<?>> fields = new ArrayList<>();
+    fields.add(
+        Fields.stringField(StoreDefinition.OperationRunsStore.NAMESPACE_FIELD, namespace));
+    fields.add(Fields.stringField(StoreDefinition.OperationRunsStore.ID_FIELD, id));
+    Long startTime = getStartTime(new OperationRunId(namespace, id));
+    fields.add(
+        Fields.longField(StoreDefinition.OperationRunsStore.START_TIME_FIELD, startTime));
+    return fields;
+  }
+
+  private OperationRunDetail<?> getCurrentRunDetail(OperationRunId runId)
```

Comment:
The generic type T refers to the type of operation request object. It is not needed for most of the cases when we retrieve the `OperationRunDetail` except when we plan to launch the operation. The current approach I am taking is based on the operation type the request object type will be determined and the OperationRunDetail will be casted accordingly.

Reply:
I'm still confused how this will work, doesn't the type need to be passed into gson in order for it to be deserialized properly? If you just deserialize it here as `?`, don't you get a Map or a JsonObject or something as the field?